agreement_rebate/models/agreement.py

```diff
@@ -0,0 +1,119 @@
+# Copyright 2020 Tecnativa - Carlos Dauden
+# Copyright 2020 Tecnativa - Sergio Teruel
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import api, models, fields
+
+
+class Agreement(models.Model):
+    _inherit = 'agreement'
+
+    def _selection_rebate_type(self):
+        return [
+            ('global', 'Global (A discount global for all lines)'),
+            ('line', 'Line (A discount for every line)'),
+            ('section_total', 'Compute total and apply discount rule match'),
+            ('section_prorated', 'Compute multi-dicounts by sections amount'),
+        ]
+
+    rebate_type = fields.Selection(
+        selection=lambda self: self._selection_rebate_type(),
+        string='rebate type',
+    )
+    rebate_line_ids = fields.One2many(
+        comodel_name='agreement.rebate.line',
+        string='Agreement rebate lines',
+        inverse_name='agreement_id',
+        copy=True,
+    )
+    rebate_section_ids = fields.One2many(
+        comodel_name='agreement.rebate.section',
+        string='Agreement rebate sections',
+        inverse_name='agreement_id',
+        copy=True,
+    )
+    rebate_discount = fields.Float(
+        string='Rebate discount',
+    )
+
+
+class AgreementRebateLine(models.Model):
+    _name = 'agreement.rebate.line'
+    _description = 'Agreement Rebate Lines'
+
+    agreement_id = fields.Many2one(
+        comodel_name='agreement',
+        string='Agreement',
+    )
+    rebate_target = fields.Selection([
+        ('product', 'Product variant'),
+        ('product_tmpl', 'Product templates'),
+        ('category', 'Product categories'),
+        ('condition', 'Rebate condition'),
+        ('domain', 'Rebate domain'),
+    ])
+    rebate_product_ids = fields.Many2many(
+        comodel_name='product.product',
+        string='Products',
+    )
+    rebate_product_tmpl_ids = fields.Many2many(
+        comodel_name='product.template',
+        string='Product templates',
+    )
+    rebate_category_ids = fields.Many2many(
+        comodel_name='product.category',
+        string='Product categories',
+    )
+    rebate_condition_id = fields.Many2one(
+        comodel_name='agreement.rebate.condition',
+        string='Rebate condition',
+    )
+    rebate_domain = fields.Char(
+        compute='_compute_rebate_domain',
+        inverse=lambda self: self,
+        string='Rebate domain',
+        store=True,
+    )
+    rebate_discount = fields.Float(string='Rebate discount')
+
+    @api.depends(
+        'rebate_target', 'rebate_product_ids', 'rebate_product_tmpl_ids',
+        'rebate_category_ids', 'rebate_condition_id')
+    def _compute_rebate_domain(self):
+        for line in self:
+            if line.rebate_target == 'product':
+                line.rebate_domain = [
```

Comment:
You are relying on the char serialization of the Python interpreter, and that can have side effects.